spring-cloud-stream-reactive/src/main/java/org/springframework/cloud/stream/reactive/StreamEmitterAnnotationBeanPostProcessor.java

```diff
@@ -0,0 +1,244 @@
+		}
+
+		if (!Void.TYPE.equals(method.getReturnType())) {
+			if (!StringUtils.hasText(outboundName)) {
+				for (int parameterIndex = 0; parameterIndex < method.getParameterTypes().length; parameterIndex++) {
+					MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method,
+							parameterIndex);
+					if (methodParameter.hasParameterAnnotation(Output.class)) {
+						outboundName = methodParameter.getParameterAnnotation(Output.class).value();
+					}
+				}
+			}
+			Object targetBean = this.applicationContext.getBean(outboundName);
+			boolean streamListenerResultAdapterFound = false;
+			for (StreamListenerResultAdapter streamListenerResultAdapter : this.streamListenerResultAdapters) {
+				if (streamListenerResultAdapter.supports(result.getClass(), targetBean.getClass())) {
+					streamListenerResultAdapter.adapt(result, targetBean);
+					streamListenerResultAdapterFound = true;
+					break;
+				}
+			}
+			Assert.isTrue(streamListenerResultAdapterFound, StreamEmitterErrorMessages.CANNOT_CONVERT_RETURN_TYPE_TO_ANY_AVAILABLE_RESULT_ADAPTERS);
+		}
+	}
+
+	private static void validateStreamEmitterMethod(Method method, int outputAnnotationCount, String methodAnnotatedOutboundName) {
+		if (StringUtils.hasText(methodAnnotatedOutboundName)) {
+			Assert.isTrue(outputAnnotationCount == 0,
+					StreamEmitterErrorMessages.INVALID_OUTPUT_METHOD_PARAMETERS);
+		}
+		else {
+			Assert.isTrue(outputAnnotationCount > 0, StreamEmitterErrorMessages.NO_OUTPUT_SPECIFIED);
+		}
+
+		if (!method.getReturnType().equals(Void.TYPE)) {
+			Assert.isTrue(StringUtils.hasText(methodAnnotatedOutboundName),
+					StreamEmitterErrorMessages.RETURN_TYPE_NO_OUTBOUND_SPECIFIED);
+			Assert.isTrue(method.getParameterCount() == 0, StreamEmitterErrorMessages.RETURN_TYPE_METHOD_ARGUMENTS);
+		}
+		else {
+			if (!StringUtils.hasText(methodAnnotatedOutboundName)) {
+				int methodArgumentsLength = method.getParameterTypes().length;
+				for (int parameterIndex = 0; parameterIndex < methodArgumentsLength; parameterIndex++) {
+					MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+					if (methodParameter.hasParameterAnnotation(Output.class)) {
+						String outboundName = (String) AnnotationUtils
+								.getValue(methodParameter.getParameterAnnotation(Output.class));
+						Assert.isTrue(StringUtils.hasText(outboundName), StreamEmitterErrorMessages.INVALID_OUTBOUND_NAME);
+					}
+					else {
+						throw new IllegalArgumentException(
+								StreamEmitterErrorMessages.OUTPUT_ANNOTATION_MISSING_ON_METHOD_PARAMETERS_VOID_RETURN_TYPE);
+					}
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return true;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		if (callback != null) {
+			callback.run();
+		}
+	}
+
+	@Override
+	public void stop() {
+		if (running) {
+			this.running = false;
```

Comment:
I am afraid we still need to implement `SmartLifeCycle` to control the ordering of this bean instantiation. We need to make sure that this bean only gets instantiated after the `OutputLifecycleBinding`. Do you think `SmartInitializingSingleton` can solve this issue?